README
This is the public website for the Inclusive Design Research Centre.


installation-old/CHANGELOG
0. Metadata
-----------
File version:	$Id: CHANGELOG 20828 2011-02-22 04:22:21Z dextercowley $
Copyright:		Copyright (C) 2005 - 2011 Open Source Matters, Inc. All rights reserved.
License:		GNU General Public License version 2 or later; see LICENSE.txt

1. Copyright and disclaimer
---------------------------
This application is opensource software released under the GPL.  Please
see source code and the LICENSE file


2. Changelog
------------
This is a non-exhaustive (but still near complete) changelog for
Joomla! 1.6, including alpha and beta versions.
Our thanks to all those people who've contributed bug reports and
code fixes.

Legend:

* -> Security Fix
# -> Issue Fix
$ -> Language fix or change
+ -> Addition
^ -> Change
- -> Removed
! -> Note

21-Feb-2011 Mark Dexter
 # [#24886] No save icon in article editor for new article with category level permission
 # [#24909] *Category Save as Copy does not save with different title/alias
 # [#22624] bug: old extension version reported from cache after automatic extension update**
 # [#24433] *modal_contacts field type problems
 # [#24963] Missing "parent" css class if menu is collapsed
 # [#24589] Systematic elimination of DS as directory separator - Round 2
 # [#24838] mod_articles_category generates a PHP warning when using language filter
 # [#24489] Joomla displaying error after uploading an image using Media Manager and the path does not stay where it is after the upload
 # [#24799] Double category strings in URLs
 # [#24524] Redirect is Not Working with SEF
 # [#24908] A banner with limited impressions will never show
 # [#24903] Cannot Translate Option Values using JForm SQL Field Type**
 # [#24857] Users with create permission in a category cannot see the image button below the editor
 # [#24993] Cannot allow a group to create in a single sub category
 # [#24307] Login form does not use https
 # [#24528] Fatal error in mod_articles_category when showing readmore
 # [#24880] Can't display teaser text in full article view.
 # [#24914] Wrong class in kategory list mode
 # [#24548] Abort during install of component and module fail due to rollback methods being protected instead of public.
 # [#24389] Plugins installed via discover are enabled
 # [#22832] PNG images are not optimized
 # [#24519] Don't use the $() function in JavaScript.
 # [#25022] Upgrade Geshi to 1.0.8.10
 # [#24675] typo in databasequery.php - udpate
 # [#25011] Beez_20 and Beez 5 xml patch
 # [#23491] Upgrade Codemirror to Version 0.94
 # [#24740] Make some scripts compatible with Mootools 1.3
 
21-Feb-2011 Jean-Marie Simonet
 # [#24999] The strings used for the display column in the module assignment slider are confusing. Thanks Elin and Andrea
 # [#24983] modifying Beez 20 to compatibility with RTL Thanks Mahmood and Ofer.
 # [#24996] There is no featured button in the tool bar, so no way to make multiple articles featured. Thanks Elin.
 # [#24929] Multiselect not implemented in all views. Thanks Ian and Andrea.
 # [#25008] JNO/JYES instead JSHOW/JHIDE in mod_weblinks.xml Thanks Serge and Bill.
 $ Adding mn-MN to installation langs
 
20-Feb-2011 Jean-Marie Simonet
 # [#24916] Missing class blog_children. Thanks Markus and Bill.
 # [#24915] Missing class default_children. Thanks Markus and Bill.
 
19-Feb-2011 Jean-Marie Simonet
 # [#24978] duplicate code weblink .php
 # [#24545] Stripped code in contents and custom html module. Thanks Amy and Janich.
 
18-Feb-2011 Jean-Marie Simonet
 # [#24975] JS patch for Beez5 - IE issue hide.js. Thanks Ole.
 # [#24947] Incorrect ID attribute's value in backend menu. Thanks Serge
 $ Deleted a duplicate string and corrected some extra spaces
 # [#24933] The "Ordering" form field type defaults to "0", causing accidental changes to ordering of plugins. Thanks Bernie and Bill.

17-Feb-2011 Mark Dexter
 # [#24231] JS error notices default template in IE7+8
 # [#24825] in sql file bug
 # [#24506] com_categories doesn't properly update paths of children categories when moved to a new parent
 # [#24532] Unused JDate code causing "Catchable fatal error"
 # [#24646] Send HTTP result code 503 for the offline page
 # [#24911] Redirect component does not list 404 errors for missing child page when parent page does exist (patch supplied)
 # [#24496] Improve installation.js
 # [#24859] Debug Language showing up
 # [#24856] Get error after save options for component
 # [#24455] Alias URL does not work with Language filter plugin active - sef off
 # [#24816] 404 errors when using pagebreak with sef enabled
 # [#24314] Remove com_search SEF encoding of search term
 # [#24896] Module Banners: "All categories" option does not include all categories
 # [#24883] Plugin User-Profile Birthday field alpha entry crashes Profile Fields in Admin
 # [#24890] A registered user can change their user name
 # [#24965] Banner Manager: Banners - increase is needed for colspan of table's footer
 # [#24419] JController class lacks a unregisterTask method

17-Feb-2011 Jean-Marie Simonet
 $ Updating ar-AA and sv-SE
 $ Adding Installation language nn-NO. Updating Credits.
 # [#24926] JTRASH instead JTRASHED in jgrid.publishedOptions. Thanks Serge.
 # [#24877] * Menu manager Multilanguage Deactivate Home
 # [#24869] Accommodating for longer string value in Banners Edit page. Thanks Bill and Andrea.

16-Feb-2011 Jean-Marie Simonet
 [#24932] *Search lower and upper limit do not take into account localise.php settings
 [#24923] *Normalise modals UI
 # [#24927] Modules do not have Trashed state. Thanks Serge.
 $ Correcting an html tag in an ini string
 #$ [#24497] Improve the voting plugin. Thanks Rouven and Christophe.
 $ Taking off underline html tag in some ini strings. Thanks Serge.
 $ Small correction in an ini file. Thanks Stella.

15-Feb-2011 Jean-Marie Simonet
 $# [#24897] Untranslated strings TPL_BEEZ5_ISCLOSED and TPL_BEEZ5_LOGO. Thanks Serge.
 # [#24884] Missing messages for some menu item types having no Basic Options. Thanks Bill.

includes/application.php
<?php
/**
 * @version		$Id: application.php 20752 2011-02-18 04:23:54Z dextercowley $
 * @copyright	Copyright (C) 2005 - 2011 Open Source Matters, Inc. All rights reserved.
 * @license		GNU General Public License version 2 or later; see LICENSE.txt
 */

// no direct access
defined('_JEXEC') or die;

jimport('joomla.application.component.helper');

/**
 * Joomla! Application class
 *
 * Provide many supporting API functions
 *
 * @package		Joomla.Site
 * @subpackage	Application
 */
final class JSite extends JApplication
{
	/**
	 * Currently active template
	 * @var object
	 */
	private $template = null;

	/**
	 * Option to filter by language
	 */
	private $_language_filter = false;

	/**
	 * Option to detect language by the browser
	 */
	private $_detect_browser = false;

	/**
	 * Class constructor
	 *
	 * @param	array An optional associative array of configuration settings.
	 * Recognized key values include 'clientId' (this list is not meant to be comprehensive).
	 */
	public function __construct($config = array())
	{
		$config['clientId'] = 0;
		parent::__construct($config);
	}

	/**
	 * Initialise the application.
	 *
	 * @param	array
	 */
	public function initialise($options = array())
	{
		$config = JFactory::getConfig();

		jimport('joomla.language.helper');

		// if a language was specified it has priority
		// otherwise use user or default language settings
		jimport('joomla.plugin.helper');
		JPluginHelper::importPlugin('system', 'languagefilter');

		if (empty($options['language'])) {
			$lang = JRequest::getString('language', null);
			if ($lang && JLanguage::exists($lang)) {
				$options['language'] = $lang;
			}
		}

		if ($this->_language_filter && empty($options['language'])) {
			// Detect cookie language
			jimport('joomla.utilities.utility');
			$lang = JRequest::getString(JUtility::getHash('language'), null ,'cookie');
			// Make sure that the user's language exists
			if ($lang && JLanguage::exists($lang)) {
				$options['language'] = $lang;
			}
		}

		if (empty($options['language'])) {
			// Detect user language
			$lang = JFactory::getUser()->getParam('language');
			// Make sure that the user's language exists
			if ($lang && JLanguage::exists($lang)) {
				$options['language'] = $lang;
			}
		}

		if ($this->_detect_browser && empty($options['language'])) {
			// Detect browser language
			$lang = JLanguageHelper::detectLanguage();
			// Make sure that the user's language exists
			if ($lang && JLanguage::exists($lang)) {
				$options['language'] = $lang;
			}
		}

		if (empty($options['language'])) {
			// Detect default language
			$params =  JComponentHelper::getParams('com_languages');
			$client	= JApplicationHelper::getClientInfo($this->getClientId());
			$options['language'] = $params->get($client->name, $config->get('language', 'en-GB'));
		}

		// One last check to make sure we have something
		if (!JLanguage::exists($options['language'])) {
			$lang = $config->get('language','en-GB');
			if (JLanguage::exists($lang)) {
				$options['language'] = $lang;
			}
			else {
				$options['language'] = 'en-GB'; // as a last ditch fail to english
			}
		}

		// Execute the parent initialise method.
		parent::initialise($options);

		// Load Library language
		$lang = JFactory::getLanguage();
		$lang->load('lib_joomla', JPATH_SITE)
		|| $lang->load('lib_joomla', JPATH_ADMINISTRATOR);

	}

	/**
	 * Route the application.
	 *
	 */
	public function route()
	{
		parent::route();

		$Itemid = JRequest::getInt('Itemid');
		$this->authorise($Itemid);
	}

	/**
	 * Dispatch the application
	 *
	 * @param	string
	 */
	public function dispatch($component = null)
	{
		// Get the component if not set.
		if (!$component) {
			$component = JRequest::getCmd('option');
		}

		$document	= JFactory::getDocument();
		$user		= JFactory::getUser();
		$router		= $this->getRouter();
		$params		= $this->getParams();

		switch($document->getType())
		{
			case 'html':
				// Get language
				$lang_code = JFactory::getLanguage()->getTag();
				$languages = JLanguageHelper::getLanguages('lang_code');

				// Set metadata
				if (isset($languages[$lang_code]) && $languages[$lang_code]->metakey) {
					$document->setMetaData('keywords', $languages[$lang_code]->metakey);
				} else {
					$document->setMetaData('keywords', $this->getCfg('MetaKeys'));
				}
				$document->setMetaData('rights', $this->getCfg('MetaRights'));
				$document->setMetaData('language', $lang_code);
				if ($router->getMode() == JROUTER_MODE_SEF) {
					$document->setBase(JURI::current());
				}
				break;

			case 'feed':
				$document->setBase(JURI::current());
				break;
		}

		$document->setTitle($params->get('page_title'));
		$document->setDescription($params->get('page_description'));
		$contents = JComponentHelper::renderComponent($component);
		$document->setBuffer($contents, 'component');

		// Trigger the onAfterDispatch event.
		JPluginHelper::importPlugin('system');
		$this->triggerEvent('onAfterDispatch');
	}

	/**
	 * Display the application.
	 */
	public function render()
	{
		$document	= JFactory::getDocument();
		$user		= JFactory::getUser();

		// get the format to render
		$format = $document->getType();

		switch ($format)
		{
			case 'feed':
				$params = array();
				break;

			case 'html':
			default:
				$template	= $this->getTemplate(true);
				$file		= JRequest::getCmd('tmpl', 'index');

				if (!$this->getCfg('offline') && ($file == 'offline')) {
					$file = 'index';
				}

				if ($this->getCfg('offline') && !$user->authorise('core.admin')) {
					$uri		= JFactory::getURI();
					$return		= (string)$uri;
					$this->setUserState('users.login.form.data',array( 'return' => $return ) );
					$file = 'offline';
					JResponse::setHeader('Status', '503 Service Temporarily Unavailable', 'true');
				}
				if (!is_dir(JPATH_THEMES.DS.$template->template) && !$this->getCfg('offline')) {
					$file = 'component';
				}
				$params = array(
					'template'	=> $template->template,
					'file'		=> $file.'.php',
					'directory'	=> JPATH_THEMES,
					'params'	=> $template->params
				);
				break;
		}

		// Parse the document.
		$document = JFactory::getDocument();
		$document->parse($params);

		// Trigger the onBeforeRender event.
		JPluginHelper::importPlugin('system');
		$this->triggerEvent('onBeforeRender');

		$caching = false;
		if ($this->getCfg('caching') && $this->getCfg('caching',2) == 2 && !$user->get('id')) {
			$caching = true;
		}

		// Render the document.
		JResponse::setBody($document->render($caching, $params));

		// Trigger the onAfterRender event.
		$this->triggerEvent('onAfterRender');
	}

	/**
	 * Login authentication function
	 *
	 * @param	array	Array('username' => string, 'password' => string)
	 * @param	array	Array('remember' => boolean)
	 *
	 * @see JApplication::login
	 */
	public function login($credentials, $options = array())
	{
		 // Set the application login entry point
		if (!array_key_exists('entry_url', $options)) {
			$options['entry_url'] = JURI::base().'index.php?option=com_users&task=user.login';
		}

		// Set the access control action to check.
		$options['action'] = 'core.login.site';

		return parent::login($credentials, $options);
	}

	/**
	 * @deprecated 1.6	Use the authorise method instead.
	 */
	public function authorize($itemid)
	{
		return $this->authorise($itemid);
	}

	/**
	 * Check if the user can access the application
	 */
	public function authorise($itemid)
	{
		$menus	= $this->getMenu();
		$user	= JFactory::getUser();

		if (!$menus->authorise($itemid))
		{
			if ($user->get('id') == 0)
			{
				// Redirect to login
				$uri		= JFactory::getURI();
				$return		= (string)$uri;

				$this->setUserState('users.login.form.data',array( 'return' => $return ) );

				$url	= 'index.php?option=com_users&view=login';
				$url	= JRoute::_($url, false);

				$this->redirect($url, JText::_('JGLOBAL_YOU_MUST_LOGIN_FIRST'));
			}
			else {
				JError::raiseError(403, JText::_('JERROR_ALERTNOAUTHOR'));
			}
		}
	}

	/**
	 * Get the appliaction parameters
	 *
	 * @param	string	The component option
	 * @return	object	The parameters object
	 * @since	1.5
	 */
	public function getParams($option = null)
	{
		static $params = array();

		$hash = '__default';
		if (!empty($option)) {
			$hash = $option;
		}
		if (!isset($params[$hash]))
		{
			// Get component parameters
			if (!$option) {
				$option = JRequest::getCmd('option');
			}
			// Get new instance of component global parameters
			$params[$hash] = clone JComponentHelper::getParams($option);

			// Get menu parameters
			$menus	= $this->getMenu();
			$menu	= $menus->getActive();

			// Get language
			$lang_code = JFactory::getLanguage()->getTag();
			$languages = JLanguageHelper::getLanguages('lang_code');

			$title = $this->getCfg('sitename');
			if (isset($languages[$lang_code]) && $languages[$lang_code]->metadesc) {
				$description = $languages[$lang_code]->metadesc;
			} else {
				$description = $this->getCfg('MetaDesc');
			}
			$rights = $this->getCfg('MetaRights');
			// Lets cascade the parameters if we have menu item parameters
			if (is_object($menu)) {
				$temp = new JRegistry;
				$temp->loadJSON($menu->params);
				$params[$hash]->merge($temp);
				$title = $menu->title;
			}

			$params[$hash]->def('page_title', $title);
			$params[$hash]->def('page_description', $description);
			$params[$hash]->def('page_rights', $rights);
		}

		return $params[$hash];
	}

	/**
	 * Get the application parameters
	 *
	 * @param	string	The component option
	 *
	 * @return	object	The parameters object
	 * @since	1.5
	 */
	public function getPageParameters($option = null)
	{
		return $this->getParams($option);
	}

	/**
	 * Get the template
	 *
	 * @return string The template name
	 * @since 1.0
	 */
	public function getTemplate($params = false)
	{
		if(is_object($this->template))
		{
			if ($params) {
				return $this->template;
			}
			return $this->template->template;
		}
		// Get the id of the active menu item
		$menu = $this->getMenu();
		$item = $menu->getActive();
		if (!$item) {
			$item = $menu->getItem(JRequest::getVar('Itemid'));
		}

		$id = 0;
		if (is_object($item)) { // valid item retrieved
			$id = $item->template_style_id;
		}
		$condition = '';

		$tid = JRequest::getVar('template', 0);
		if (is_int($tid) && $tid > 0) {
			$id = (int) $tid;
		}


		$cache = JFactory::getCache('com_templates', '');
		if ($this->_language_filter) {
			$tag = JFactory::getLanguage()->getTag();
		}
		else {
			$tag ='';
		}
		if (!$templates = $cache->get('templates0'.$tag)) {
			// Load styles
			$db = JFactory::getDbo();
			$query = $db->getQuery(true);
			$query->select('id, home, template, params');
			$query->from('#__template_styles');
			$query->where('client_id = 0');

			$db->setQuery($query);
			$templates = $db->loadObjectList('id');
			foreach($templates as &$template) {
				$registry = new JRegistry;
				$registry->loadJSON($template->params);
				$template->params = $registry;

				// Create home element
				if ($template->home == '1' && !isset($templates[0]) || $this->_language_filter && $template->home == $tag) {
					$templates[0] = clone $template;
				}
			}
			$cache->store($templates, 'templates0'.$tag);
		}

		$template = $templates[$id];

		// Allows for overriding the active template from the request
		$template->template = JRequest::getCmd('template', $template->template);
		$template->template = JFilterInput::getInstance()->clean($template->template, 'cmd'); // need to filter the default value as well

		// Fallback template
		if (!file_exists(JPATH_THEMES.DS.$template->template.DS.'index.php')) {
			JError::raiseWarning(0, JText::_('JERROR_ALERTNOTEMPLATE'));
		    $template->template = 'beez_20';
		    if (!file_exists(JPATH_THEMES.DS.'beez_20'.DS.'index.php')) {
		    	$template->template = '';
		    }
		}

		// Cache the result
		$this->template = $template;
		if ($params) {
			return $template;
		}
		return $template->template;
	}

	/**
	 * Overrides the default template that would be used
	 *
	 * @param string The template name
	 */
	public function setTemplate($template)
	{
		if (is_dir(JPATH_THEMES.DS.$template)) {
			$this->template = new stdClass();
			$this->template->params = new JRegistry;
			$this->template->template = $template;
		}
	}

	/**
	 * Return a reference to the JPathway object.
	 *
	 * @param	string	$name		The name of the application/client.
	 * @param	array	$options	An optional associative array of configuration settings.
	 *
	 * @return	object	JMenu.
	 * @since	1.5
	 */
	public function getMenu($name = null, $options = array())
	{
		$options	= array();
		$menu		= parent::getMenu('site', $options);
		return $menu;
	}

	/**
	 * Return a reference to the JPathway object.
	 *
	 * @param	string	$name		The name of the application.
	 * @param	array	$options	An optional associative array of configuration settings.
	 *
	 * @return	object JPathway.
	 * @since	1.5
	 */
	public function getPathway($name = null, $options = array())
	{
		$options = array();
		$pathway = parent::getPathway('site', $options);
		return $pathway;
	}

	/**
	 * Return a reference to the JRouter object.
	 *
	 * @param	string	$name		The name of the application.
	 * @param	array	$options	An optional associative array of configuration settings.
	 *
	 * @return	JRouter
	 * @since	1.5
	 */
	static public function getRouter($name = null, array $options = array())
	{
		$config = JFactory::getConfig();
		$options['mode'] = $config->get('sef');
		$router = parent::getRouter('site', $options);
		return $router;
	}

	/**
	 * Return the current state of the language filter.
	 *
	 * @return	boolean
	 * @since	1.6
	 */
	public function getLanguageFilter()
	{
		return $this->_language_filter;
	}

	/**
	 * Set the current state of the language filter.
	 *
	 * @return	boolean	The old state
	 * @since	1.6
	 */
	public function setLanguageFilter($state=false)
	{
		$old = $this->_language_filter;
		$this->_language_filter=$state;
		return $old;
	}
	/**
	 * Return the current state of the detect browser option.
	 *
	 * @return	boolean
	 * @since	1.6
	 */
	public function getDetectBrowser()
	{
		return $this->_detect_browser;
	}

	/**
	 * Set the current state of the detect browser option.
	 *
	 * @return	boolean	The old state
	 * @since	1.6
	 */
	public function setDetectBrowser($state=false)
	{
		$old = $this->_detect_browser;
		$this->_detect_browser=$state;
		return $old;
	}

	/**
	 * Redirect to another URL.
	 *
	 * Optionally enqueues a message in the system message queue (which will be displayed
	 * the next time a page is loaded) using the enqueueMessage method. If the headers have
	 * not been sent the redirect will be accomplished using a "301 Moved Permanently"
	 * code in the header pointing to the new location. If the headers have already been
	 * sent this will be accomplished using a JavaScript statement.
	 *
	 * @param	string	The URL to redirect to. Can only be http/https URL
	 * @param	string	An optional message to display on redirect.
	 * @param	string  An optional message type.
	 * @param	boolean	True if the page is 301 Permanently Moved, otherwise 303 See Other is assumed.
	 * @param	boolean	True if the enqueued messages are passed to the redirection, false else.
	 * @return	none; calls exit().
	 * @since	1.5
	 * @see		JApplication::enqueueMessage()
	 */
	public function redirect($url, $msg='', $msgType='message', $moved = false, $persistMsg = true)
	{
		if (!$persistMsg) {
			$this->_messageQueue = array();
		}
		parent::redirect($url, $msg, $msgType, $moved);
	}
}
